### DISCUSSION TITLE
Installing a Visual Studio Extension (.VSIX) from the Marketplace in a Windows runner?

### DISCUSSION THREAD

**Discussion by gbakeman** [community/community#49461] [Actions] [answered]
### Select Topic Area

Question

### Body

My project uses the [Multilingual App Toolkit](https://marketplace.visualstudio.com/items?itemName=dts-publisher.mat2022) extension from Microsoft in order to properly compile languages into my app. I'm slowly building up my CI/CD pipeline so the entire build process is automated, although I've hit a roadblock with respect to this extension.

The problem is, I need to install this Visual Studio extension from the Extensions Marketplace in order for the language files to be properly compiled. Doing that doesn't appear to be straightforward. I've located an [install-vsix](https://github.com/microcompiler/install-vsix) Action that someone created, which is based off of a [Powershell script](https://gist.github.com/ScottHutchinson/b22339c3d3688da5c9b477281e258400) by someone else. The script will hang at the Installation stage however, and based on the testing that I've done on my own system, I'm assuming this is because the extension installer is trying to request admin rights from a user who isn't present.

A GitHub issue [suggests](https://github.com/actions/runner/issues/408) that we configure the runner service to run with `SYSTEM` permissions, but I'm not sure if this is the officially recommended solution or even a safe/good idea.

**Comment by 0xNevo** [comment]
Yeah, Of course.

$vsixUrl = "https://marketplace.visualstudio.com/_apis/public/gallery/publishers/Microsoft/vsextensions/MultilingualAppToolkit.VSIX/4.0.1579.0/vspackage"
$vsixPath = "$env:TEMP\MultilingualAppToolkit.vsix"
Invoke-WebRequest -Uri $vsixUrl -OutFile $vsixPath
& "$env:ProgramFiles (x86)\Microsoft Visual Studio\2019\Enterprise\Common7\IDE\VSIXInstaller.exe" $vsixPath /q

**Comment by gbakeman** [comment]
Hi @SayedAlAhlMohamed and @comet19950902 ,

Thank you both for your responses. I just checked, and unfortunately the script that I was using already calls the VSIXInstaller with the /q switch. I think the UAC prompt is independent of that anyways. Am I to understand then that the only solution at this point is to elevate the runner service to the `SYSTEM` permission level?

**Reply by 0xNevo** [reply]
Yes, that might be the only solution left to handle the UAC prompt. Elevating the runner service to run with SYSTEM permission level will allow it to bypass the UAC prompt that requires user approval before proceeding with the installer.

However, elevating the runner service may have security implications and is not recommended unless absolutely necessary. You should consider the potential risks and evaluate whether the benefits outweigh them before taking this step.

Alternatively, you can also try using a different installer package that does not raise the UAC prompt or disable the UAC prompt in your system temporarily during the installation process. But remember to enable it back after installation to maintain system security.

**Reply by gbakeman** [reply]
Since we're discussing using a GitHub-hosted runner here (Windows Server), does that change the situation at all? Is there some way to disable the UAC system from a workflow, or do I even need to worry that much about elevating the runner service to a higher permission level?

**Reply by 0xNevo** [reply]
If you're running a GitHub-hosted runner on Windows Server, the User Account Control (UAC) system is still in place, so it may affect your workflows that require elevated permissions.

To bypass UAC from a workflow, you'll need to modify the elevate.exe command, which is used to run commands with administrative privileges on Windows. You can add the /c argument followed by your command to the end of the elevate.exe command to run the specified command as an administrator without prompting for elevation.

Here's an example of how to use elevate.exe to run a PowerShell script with administrative permissions:

  name: Run PowerShell script with elevated permissions
  shell: powershell
  run: |
    elevate.exe /c powershell.exe -ExecutionPolicy Bypass -File your-script.ps1

This will execute your PowerShell script with elevated privileges without receiving any prompts or user intervention.

Additionally, it's important to note that you shouldn't worry about elevating the runner service to a higher permission level as GitHub-hosted runners are already configured with appropriate permissions and access levels by default.

Overall, depending on the specifics of your workflow, disabling UAC may not be necessary, but using the elevate.exe command will allow you to run commands with elevated permissions when needed.

**Comment by gbakeman** [comment, marked as answer]
Update: So either something has changed that I'm not aware of, or I simply wasn't waiting long enough: It turns out installing this extension on the runner machine was just taking a very long time (~20 minutes) compared to doing it on my own machine. Thanks everyone!
